contracts/lens/Quoter.sol

```diff
@@ -0,0 +1,270 @@
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+import {PoolManager} from "@uniswap/v4-core/contracts/PoolManager.sol";
+
+contract Quoter is IQuoter {
+    using PoolIdLibrary for PoolKey;
+    using SafeCast for *;
+    using Hooks for IHooks;
+
+    // v4 Singleton contract
+    IPoolManager poolManager;
+
+    constructor(address _poolManager) {
+        poolManager = IPoolManager(_poolManager);
+    }
+
+    function fillSlot0(PoolId id) private view returns (Pool.Slot0 memory slot0) {
+        //TODO: extsload when storage is stable?
+        (slot0.sqrtPriceX96, slot0.tick,,) = poolManager.getSlot0(id);
+
+        return slot0;
+    }
+
+    function validateRevertReason(bytes memory reason) private pure returns (bytes memory) {
+        if (reason.length < 96) {
+            // function selector + length of bytes as uint256 + min length of revert reason padded to multiple of 32 bytes
+            if (reason.length < 68) {
+                revert UnexpectedRevertBytes();
+            }
+            assembly {
+                reason := add(reason, 0x04)
+            }
+            revert(abi.decode(reason, (string)));
+        }
+        return reason;
+    }
+
+    function _handleRevertExactInputSingle(bytes memory reason, PoolKey memory poolKey)
+        private
+        view
+        returns (int128[] memory deltaAmounts, uint160 sqrtPriceX96After, uint32 initializedTicksCrossed)
+    {
+        int24 tickBefore;
+        int24 tickAfter;
+        BalanceDelta deltas;
+        deltaAmounts = new int128[](2);
+        (, tickBefore,,) = poolManager.getSlot0(poolKey.toId());
+        reason = validateRevertReason(reason);
+        (deltas, sqrtPriceX96After, tickAfter) = abi.decode(reason, (BalanceDelta, uint160, int24));
+        deltaAmounts[0] = deltas.amount0();
+        deltaAmounts[1] = deltas.amount1();
+
+        initializedTicksCrossed =
+            PoolTicksCounter.countInitializedTicksCrossed(poolManager, poolKey, tickBefore, tickAfter);
+    }
+
+    function _handleRevertExactInput(bytes memory reason)
+        private
+        pure
+        returns (
+            int128[] memory deltaAmounts,
+            uint160[] memory sqrtPriceX96AfterList,
+            uint32[] memory initializedTicksCrossedList
+        )
+    {
+        reason = validateRevertReason(reason);
+        (deltaAmounts, sqrtPriceX96AfterList, initializedTicksCrossedList) =
+            abi.decode(reason, (int128[], uint160[], uint32[]));
+    }
+
+    function lockAcquired(bytes calldata encodedSwapIntention) external returns (bytes memory) {
+        require(msg.sender == address(poolManager));
+
+        SwapInfo memory swapInfo = abi.decode(encodedSwapIntention, (SwapInfo));
+
+        if (swapInfo.swapType == SwapType.ExactInputSingle) {
+            (BalanceDelta deltas, uint160 sqrtPriceX96After, int24 tickAfter) =
```

Comment:
think each branch here deserves its own helper function

Reply:
one nice pattern that we've used in the past when lock has multiple branches is to encode a sub-call in the lock and self-call in lockAcquired
as defined here https://github.com/Uniswap/v4-periphery/blob/581d96dfd8b281cafe351205bf6d2d65efb4df90/contracts/BaseHook.sol#L52-L61

Reply:
Still think it's worth looking into the self call pattern here for cleaner branching